src/demangle.cpp

```diff
@@ -25,80 +25,23 @@ char* Demangle::demangleCpp(const char* s) {
     return result;
 }
 
-char* Demangle::demangleRust(const char* s, const char* e) {
+char* Demangle::demangleRust(const char *s, struct demangle const *demangle, bool full_signature) {
     // Demangled symbol can be 1.5x longer than original, e.g. 1A1B1C -> A::B::C
-    char* result = (char*)malloc((e - s) * 3 / 2 + 1);
-    if (result == NULL) {
-        return NULL;
-    }
-
-    char* r = result;
-    char* tmp;
-
-    while (s < e) {
-        unsigned long len = strtoul(s, &tmp, 10);
-        const char* next = tmp + len;
-        if (len == 0 || next > e) {
-            break;
-        }
-
-        s = tmp;
-        if (s[0] == '_' && s[1] == '$') s++;
-
-        if (r > result) {
-            *r++ = ':';
-            *r++ = ':';
+    char* result = NULL, *new_result;
+    for (size_t demangled_size = strlen(s) * 3 / 2 + 1; demangled_size < 1000000; demangled_size *= 2) {
+        new_result = (char *)realloc(result, demangled_size);
```

Comment:
`realloc` is not used for its intended purpose (no need to copy previous data).
The code will be even simpler if we use `malloc` here and `free` after `rust_demangle_display_demangle`.